In [1]:
!pip install pandas

In [3]:
#this code chunk should be run before running the code for any of the individual questions
import pandas as pd
import numpy as np
intakes = pd.read_csv('data/aac_intakes.csv')
outcomes = pd.read_csv('data/aac_outcomes.csv')
#This code is required to run the code for two or more questions
intakes['year_found'] = pd.DatetimeIndex(intakes['datetime']).year
intakes['month_found'] = pd.DatetimeIndex(intakes['datetime']).month
adopted_pets = outcomes.loc[outcomes['outcome_type']=='Adoption']


In [2]:
intakes.head(1)

age_upon_intake animal_id animal_type                     breed  \
0         8 years   A706918         Dog  English Springer Spaniel   

         color                 datetime                datetime2  \
0  White/Liver  2015-07-05T12:59:00.000  2015-07-05T12:59:00.000   

                     found_location intake_condition intake_type   name  \
0  9409 Bluegrass Dr in Austin (TX)           Normal       Stray  Belle   

  sex_upon_intake  
0   Spayed Female

In [3]:
outcomes.head(1)

age_upon_outcome animal_id animal_type                   breed  \
0          2 weeks   A684346         Cat  Domestic Shorthair Mix   

          color        date_of_birth             datetime  \
0  Orange Tabby  2014-07-07T00:00:00  2014-07-22T16:04:00   

             monthyear name outcome_subtype outcome_type sex_upon_outcome  
0  2014-07-22T16:04:00  NaN         Partner     Transfer      Intact Male

In [4]:
intakes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80187 entries, 0 to 80186
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   age_upon_intake   80187 non-null  object
 1   animal_id         80187 non-null  object
 2   animal_type       80187 non-null  object
 3   breed             80187 non-null  object
 4   color             80187 non-null  object
 5   datetime          80187 non-null  object
 6   datetime2         80187 non-null  object
 7   found_location    80187 non-null  object
 8   intake_condition  80187 non-null  object
 9   intake_type       80187 non-null  object
 10  name              55603 non-null  object
 11  sex_upon_intake   80186 non-null  object
dtypes: object(12)
memory usage: 7.3+ MB


In [18]:
outcomes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80681 entries, 0 to 80680
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   age_upon_outcome  80673 non-null  object
 1   animal_id         80681 non-null  object
 2   animal_type       80681 non-null  object
 3   breed             80681 non-null  object
 4   color             80681 non-null  object
 5   date_of_birth     80681 non-null  object
 6   datetime          80681 non-null  object
 7   monthyear         80681 non-null  object
 8   name              56116 non-null  object
 9   outcome_subtype   36893 non-null  object
 10  outcome_type      80667 non-null  object
 11  sex_upon_outcome  80679 non-null  object
dtypes: object(12)
memory usage: 7.4+ MB


**1) Top Five Places Where Animals Are Found**

In [14]:
intakes.groupby('found_location').size().sort_values(ascending=False)[:5]

found_location
Austin (TX)                          14443
Outside Jurisdiction                   948
Travis (TX)                            921
7201 Levander Loop in Austin (TX)      517
Del Valle (TX)                         411
dtype: int64

I got this answer by grouping the values by the 'found_location' variable, counting the number of rows for each grouping, sorting the counts in descending order and then finding the first five results.

The answer is Austin (TX), Outside Jurisdiction, Travis(TX), 7201 Levander Loop in Austin (TX), and Del Valle (TX)

**2) What is the average number of pets found in a month in the year 2015? Are there months where there is a higher number of animals found?**

In [7]:
intakes['year_found'] = pd.DatetimeIndex(intakes['datetime']).year

In [8]:
intakes['month_found'] = pd.DatetimeIndex(intakes['datetime']).month

In [41]:
intakes_by_month = intakes.loc[intakes['year_found']==2015].groupby('month_found').size()

In [37]:
intakes_by_month.mean()

1559.3333333333333

In [43]:
intakes_by_month > 1560

month_found
1     False
2     False
3     False
4     False
5      True
6      True
7      True
8      True
9      True
10     True
11    False
12    False
dtype: bool

**The mean number of intakes per month in 2015 is 1559.3333333333333 or 1560 pets. The months where the number of intakes were greater than this average were May to October 2015. I found these answers by creating a new data frame called intakes_by_month that had the number of intakes per month in 2015. I then found the average of these counts using the mean() function and then used a comparison operator to figure out which months had counts that were greater than 1560.** 

**3) Ratio of incoming pets versus adopted pets** 

In [24]:
len(intakes)

80187

This helps you describe the unique values in a certain column

In [10]:
outcomes.outcome_type.unique()


array(['Transfer', 'Adoption', 'Euthanasia', 'Return to Owner', 'Died',
       'Disposal', 'Relocate', 'Missing', nan, 'Rto-Adopt'], dtype=object)

In [5]:
adopted_pets = outcomes.loc[outcomes['outcome_type']=='Adoption']

In [21]:
len(adopted_pets)

34232

In [25]:
80187/34232

2.3424573498480954

**The ratio of incoming pets to adopted pets is 2.34. I calculated this by first finding the number of pets in the incoming database and dividing that number by the number of pets from the outcomes database that were classified as having an outcome type - 'Adoption'. Then I divided these two numbers to find the ratio.**

**4) Type of Each Animal in the Shelter** This is the count of each type of animal in the shelter.

In [6]:
intakes.groupby('animal_type').size()

animal_type
Bird           342
Cat          29659
Dog          45743
Livestock        9
Other         4434
dtype: int64

The answer above shows the number of each type of animal in the shelter. I found this by using the pandas size()and groupby() functions. 

**5)What are the adoption rates for specific breeds?
Find the top 5 dog breeds in the shelter (based on count) and then find the adoption percentage of each breed.**

In [19]:
breeds_series = intakes.loc[intakes['animal_type']=='Dog'].groupby('breed').size().sort_values(ascending=False)[:5]

In [20]:
breeds_frame = pd.DataFrame(breeds_series)

In [40]:
breeds_frame = breeds_frame.rename(columns={0: 'Count'})

In [41]:
# Calculate Percentage
breeds_frame['adoption_percent'] = (breeds_frame['Count'] / 
                      adopted_pets['breed'].count()) * 100
# Show the dataframe
breeds_frame

Count  adoption_percent
breed                                             
Pit Bull Mix                6382         18.643375
Chihuahua Shorthair Mix     4860         14.197242
Labrador Retriever Mix      4841         14.141739
German Shepherd Mix         1963          5.734401
Australian Cattle Dog Mix   1105          3.227974

**The table above shows the adoption percentage of each breed. To find this result I filtered out the dogs in the intakes database, grouped the results by breed and sorted the data by number of intakes per breed. Then I put this series into a dataframe so that I could better manipulate the data. After doing this I divided the number of intakes per breed by the total number of adopted pets in the shelter and multiplied the result by 100**


**6) What are the adoption rates for different colorings?**
Find the top 5 colorings in the shelter (based on count) and then find the adoption percentage of each color.

In [10]:
colors_series = intakes.groupby('color').size().sort_values(ascending=False)[:5]
colors_frame = pd.DataFrame(colors_series)
colors_frame = colors_frame.rename(columns={0: 'Count'})
# Calculate Percentage
colors_frame['adoption_percent'] = (colors_frame['Count'] / adopted_pets['color'].count()) * 100
# Show the dataframe
colors_frame

Count  adoption_percent
color                               
Black/White   8340         24.363169
Black         6710         19.601542
Brown Tabby   4487         13.107619
Brown         3618         10.569058
White         2849          8.322622

**The table above contains the answer to the question. I found this in a similar way to how I found the answer to the previous question except I did not include a filter to only analyze the dogs in the database.**

**7)About how many animals are spayed/neutered each month?**
This will help the shelter allocate resources and staff. Assume that all intact males and females will be spayed/neutered.

In [6]:
# Filters intakes by the intact females and males
intact = intakes.loc[((intakes['sex_upon_intake']=='Intact Female') | (intakes['sex_upon_intake']=='Intact Male' ))]

In [13]:
# N.B that 'table' is a DataFrame
table = pd.pivot_table(intact, values='animal_id', index=['month_found'],
                    columns=['year_found'], aggfunc='count')
#Finds the average number of intakes per month over the years
table['Average'] = table.mean(axis = 1)
# Rename the months with month names instead of numbers
table = table.rename(index={1: 'January', 2: 'February', 3: 'March', 4: 'April', 5: 'May', 6: 'June'})
table = table.rename(index={7: 'July', 8: 'August', 9: 'September', 10: 'October', 11: 'November', 12: 'December'})
table.round()


year_found     2013    2014    2015    2016    2017   2018  Average
month_found                                                        
January         NaN   755.0   718.0   712.0   714.0  651.0    710.0
February        NaN   685.0   611.0   674.0   650.0  613.0    647.0
March           NaN   815.0   699.0   708.0   711.0  613.0    709.0
April           NaN   958.0   809.0   870.0   898.0    NaN    884.0
May             NaN  1384.0  1293.0  1296.0  1241.0    NaN   1304.0
June            NaN  1223.0  1438.0  1072.0  1275.0    NaN   1252.0
July            NaN  1334.0  1046.0   874.0  1044.0    NaN   1074.0
August          NaN  1084.0  1061.0  1046.0   941.0    NaN   1033.0
September       NaN  1136.0   994.0   955.0  1111.0    NaN   1049.0
October      1079.0  1023.0  1072.0   830.0   869.0    NaN    975.0
November      833.0   856.0   865.0   829.0   724.0    NaN    821.0
December      767.0   752.0   649.0   823.0   665.0    NaN    731.0

**The table above contains the average number of animals that were spayed/neutered per month in the right-most column. I found this result by first creating a pivot table to sort the number of animals brought into the shelter for each month and year(I had already added 2 additional columns to the intake data set with the year and the month that each animal was found). After, I added another column titled "Average" that calculated the average across each row to show the average number of animals that were brought into the shelter for a particular month over the years.**

**Exceeds Question 1)
How many animals in the shelter are repeats? Which animal was returned to the shelter the most?**
This means the animal has been brought in more than once.

In [16]:
#Number of animals in the shelter who are repeats
intakes.duplicated(subset ='animal_id').sum()

7822

In [53]:
intakes.groupby('animal_id').animal_id.count().sort_values(ascending=False)

animal_id
A721033    13
A718223    11
A706536    11
A716018     9
A694501     8
           ..
A697715     1
A697717     1
A697721     1
A697722     1
A769067     1
Name: animal_id, Length: 72365, dtype: int64